Subject: [PATCH] Provide synchronization in doTrimForeground to avoid NPE

If doDie() is not called from main thread, then mView property
of a View object may have inconsistent value. For instance,
dispatchDetachedFromWindow() may set mView to null,
causing NullPointerException.

diff --git a/core/java/android/view/ViewRootImpl.java b/core/java/android/view/ViewRootImpl.java
index 7ace841..4bab567 100644
--- a/core/java/android/view/ViewRootImpl.java
+++ b/core/java/android/view/ViewRootImpl.java
@@ -626,6 +626,17 @@
         }
     }
 
+    public boolean doTrimForeground() {
+        synchronized(this) {
+            if (mView != null && getHostVisibility() == View.VISIBLE
+                    && mAttachInfo.mThreadedRenderer != null) {
+                return true;
+            } else {
+                destroyHardwareResources();
+            }
+            return false;
+        }
+    }
     /**
      * We have one child
      */
diff --git a/core/java/android/view/WindowManagerGlobal.java b/core/java/android/view/WindowManagerGlobal.java
index c7e8dee..6d3c5ec 100644
--- a/core/java/android/view/WindowManagerGlobal.java
+++ b/core/java/android/view/WindowManagerGlobal.java
@@ -534,12 +534,7 @@
         synchronized (mLock) {
             for (int i = mRoots.size() - 1; i >= 0; --i) {
                 final ViewRootImpl root = mRoots.get(i);
-                if (root.mView != null && root.getHostVisibility() == View.VISIBLE
-                        && root.mAttachInfo.mThreadedRenderer != null) {
-                    hasVisibleWindows = true;
-                } else {
-                    root.destroyHardwareResources();
-                }
+                hasVisibleWindows |= root.doTrimForeground();
             }
         }
         if (!hasVisibleWindows) {
